revocation/revocation.go

```diff
@@ -16,49 +16,97 @@
 package revocation
 
 import (
+	"context"
 	"crypto/x509"
 	"errors"
+	"fmt"
 	"net/http"
 	"time"
 
 	"github.com/notaryproject/notation-core-go/revocation/ocsp"
 	"github.com/notaryproject/notation-core-go/revocation/result"
 )
 
-// Revocation is an interface that specifies methods used for revocation checking
+// Revocation is an interface that specifies methods used for revocation checking.
+//
+// Deprecated: Revocation exists for backwards compatibility and should not be used.
+// To perform revocation check, use [ContextRevocation].
 type Revocation interface {
 	// Validate checks the revocation status for a certificate chain using OCSP
 	// and returns an array of CertRevocationResults that contain the results
 	// and any errors that are encountered during the process
 	Validate(certChain []*x509.Certificate, signingTime time.Time) ([]*result.CertRevocationResult, error)
 }
 
+// ValidateContextOptions provides configuration options for revocation checks
+type ValidateContextOptions struct {
+	// CertChain denotes the certificate chain whose revocation status is
+	// been validated. REQUIRED.
+	CertChain []*x509.Certificate
+
+	// AuthenticSigningTime denotes the authentic signing time of the signature.
+	// It is solely used under signing scheme `notary.x509.signingAuthority`.
+	// OPTIONAL.
+	AuthenticSigningTime time.Time
```

Comment:
For this package signingTime has nothing to do with `notary.x509.signingAuthority`, its used for revocation check using CRL to determine if signature was generated before certificate was compromised.

Reply:
This PR is about `revocation` and `ocsp`. 
According to the Notary Project `ocsp` spec: https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#ocsp section `3.i.a` states that
```
If the invalidity date is present and the authentic signing time is also present then if the invalidity date is before the authentic signing time, the certificate is considered revoked.
```
Therefore, it should be `AuthenticSigningTime` and it's a valid concept solely under `notary.x509.signingAuthority`. For `notary.x509`, we are passing zero value here because it only supports RFC 3161 timestamp.

Reply:
The word `authentic signing time` is misleading its been used to refer to anytimestamping source that user can trust like TSA in case of notary.x509 or  notary.x509.signingAuthority

Reply:
Okay, let me change it back to `SigningTime` for now. 

Created an issue in the Notary Project spec repo to track this topic: https://github.com/notaryproject/specifications/issues/309

Reply:
to confirm, we are passing signing time when signature is timestamped using TSA?

Reply:
In notation-go, when signing scheme is `notary.x509`, we are passing the zero value, i.e., `time.Time{}` into signing time. In this way, we always say the cert is revoked, `If the certificate is revoked i.e CertStatus is revoked`

Reply:
With time-stamping we do have autentic signing time.

Reply:
`Timestamping` is a different concept than `Authentic signing time`, we should not mix them. Spec: https://github.com/notaryproject/specifications/blob/main/specs/signature-specification.md#signing-time--authentic-signing-time

Based on above, we'd better keep the name `SigningTime` in this PR, as it's more general.

Reply:
At core they provide similar value. when we know (using timestamping) signature was done before some time, we can use that time and compare it with `InvalidityDate` of CRL to know its revocation status.

Reply:
@priteshbandi Changed back to `AuthenticSigningTime`.